Dockerfile-linux.template

```diff
@@ -0,0 +1,460 @@
+# Enable optimization (-O2)
+# Enable linker optimization (this sorts the hash buckets to improve cache locality, and is non-default)
+# https://github.com/docker-library/php/issues/272
+# -D_LARGEFILE_SOURCE and -D_FILE_OFFSET_BITS=64 (https://www.php.net/manual/en/intro.filesystem.php)
+ENV PHP_CFLAGS="-fstack-protector-strong -fpic -fpie -O2 -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64"
+ENV PHP_CPPFLAGS="$PHP_CFLAGS"
+ENV PHP_LDFLAGS="-Wl,-O1 -pie"
+
+ENV GPG_KEYS {{ .gpgKeys }}
+
+ENV PHP_VERSION {{ .version }}
+ENV PHP_URL="{{ .url }}" PHP_ASC_URL="{{ .ascUrl // "" }}"
+ENV PHP_SHA256="{{ .sha256 // "" }}"
+
+RUN set -eux; \
+	\
+{{ if is_alpine then ( -}}
+	apk add --no-cache --virtual .fetch-deps gnupg; \
+{{ ) else ( -}}
+	savedAptMark="$(apt-mark showmanual)"; \
+	apt-get update; \
+	apt-get install -y --no-install-recommends gnupg dirmngr; \
+	rm -rf /var/lib/apt/lists/*; \
+{{ ) end -}}
+	\
+	mkdir -p /usr/src; \
+	cd /usr/src; \
+	\
+	curl -fsSL -o php.tar.xz "$PHP_URL"; \
+	\
+	if [ -n "$PHP_SHA256" ]; then \
+		echo "$PHP_SHA256 *php.tar.xz" | sha256sum -c -; \
+	fi; \
+	\
+	if [ -n "$PHP_ASC_URL" ]; then \
+		curl -fsSL -o php.tar.xz.asc "$PHP_ASC_URL"; \
+		export GNUPGHOME="$(mktemp -d)"; \
+		for key in $GPG_KEYS; do \
+			gpg --batch --keyserver keyserver.ubuntu.com --recv-keys "$key"; \
+		done; \
+		gpg --batch --verify php.tar.xz.asc php.tar.xz; \
+		gpgconf --kill all; \
+		rm -rf "$GNUPGHOME"; \
+	fi; \
+	\
+{{ if is_alpine then ( -}}
+	apk del --no-network .fetch-deps
+{{ ) else ( -}}
+	apt-mark auto '.*' > /dev/null; \
+	apt-mark manual $savedAptMark > /dev/null; \
+	apt-get purge -y --auto-remove -o APT::AutoRemove::RecommendsImportant=false
+{{ ) end -}}
+
+COPY docker-php-source /usr/local/bin/
+
+RUN set -eux; \
+{{ if is_alpine then ( -}}
+	apk add --no-cache --virtual .build-deps \
+{{ ) else ( -}}
+	\
+	savedAptMark="$(apt-mark showmanual)"; \
+	apt-get update; \
+	apt-get install -y --no-install-recommends \
+{{ ) end -}}
+{{
+	[
+		[
+			# common packages
+			"libedit-dev",
+			"libxml2-dev",
+			( if (.version | version_id) >= ("7.2" | version_id) then "libsodium-dev" else empty end ),
+	
+			( if is_alpine then
+				"$PHPIZE_DEPS",
+					"argon2-dev",
+				"coreutils", 
+				"curl-dev",
+				"openssl-dev",
+				"sqlite-dev",
+				# https://github.com/docker-library/php/issues/888
+				( if (.version | version_id) >= ("7.4" | version_id) then "linux-headers" else empty end ),
+				# oniguruma is part of mbstring in php 7.4+
+				( if (.version | version_id) >= ("7.4" | version_id) then "oniguruma-dev" else empty end )
+			else
+				# debian packages
+				"${PHP_EXTRA_BUILD_DEPS:-}",
+				"libargon2-dev",
+				"libcurl4-openssl-dev",
+				"libsqlite3-dev",
+				"libssl-dev",
+				"zlib1g-dev",
+				# oniguruma is part of mbstring in php 7.4+
+				( if (.version | version_id) >= ("7.4" | version_id) then "libonig-dev" else empty end )
+			end )
+		] | sort[] | (
```

Comment:
This block is very clever, but IMO will be more complex to maintain over time than just having the package list, especially since there are only three packages who actually share the same name between Debian and Alpine (and we have to duplicate the version checks anyways). :grimacing:

Reply:
I don't mind moving the "common" packages into both alpine and debian lists.  What I like about it is that the list in the resulting Dockerfile is always sorted without having to manually sort the list along with  `{{ if .version then ( -}}` packages.

Reply:
To put it clearly, do we want to have maintain the sort on this:

```Dockerfile
{{ if is_alpine then ( -}}
	apk add --no-cache --virtual .build-deps \
		$PHPIZE_DEPS \
		argon2-dev \
		coreutils \ 
		curl-dev \
		libedit-dev \
{{ if (.version | version_id) >= ("7.2" | version_id) then ( -}}
		libsodium-dev \
{{ ) else "" end -}}
		libxml2-dev \
{{ if (.version | version_id) >= ("7.4" | version_id) then ( -}}
		linux-headers \
{{ ) else "" end -}}
{{ if (.version | version_id) >= ("7.4" | version_id) then ( -}}
		# oniguruma is part of mbstring in php 7.4+
		oniguruma-dev \
{{ ) else "" end -}}
		openssl-dev \
		sqlite-dev \
		# https://github.com/docker-library/php/issues/888
{{ ) else ( -}}
	\
	savedAptMark="$(apt-mark showmanual)"; \
	apt-get update; \
	apt-get install -y --no-install-recommends \
		${PHP_EXTRA_BUILD_DEPS:-} \
		libargon2-dev \
		libcurl4-openssl-dev \
		libedit-dev \
{{ if (.version | version_id) >= ("7.4" | version_id then ( -}}
		# oniguruma is part of mbstring in php 7.4+
		libonig-dev \
{{ ) else "" end -}}
{{ if (.version | version_id) >= ("7.2" | version_id) then ( -}}
		libsodium-dev \
{{ ) else "" end -}}
		libsqlite3-dev \
		libssl-dev \
		libxml2-dev \
		zlib1g-dev \
{{ ) end -}}
```

Or this (maybe some love to reduce indentation):
```jq
{{
	[
		[
			( if is_alpine then
				"$PHPIZE_DEPS",
				"argon2-dev",
				"coreutils", 
				"curl-dev",
				"libedit-dev",
				"libxml2-dev",
				"openssl-dev",
				"sqlite-dev",
				( if (.version | version_id) >= ("7.2" | version_id) then "libsodium-dev" else empty end ),
				# https://github.com/docker-library/php/issues/888
				( if (.version | version_id) >= ("7.4" | version_id) then "linux-headers" else empty end ),
				# oniguruma is part of mbstring in php 7.4+
				( if (.version | version_id) >= ("7.4" | version_id) then "oniguruma-dev" else empty end )
			else
				# debian packages
				"${PHP_EXTRA_BUILD_DEPS:-}",
				"libargon2-dev",
				"libcurl4-openssl-dev",
				"libedit-dev",
				"libsqlite3-dev",
				"libssl-dev",
				"libxml2-dev",,
				"zlib1g-dev"
				( if (.version | version_id) >= ("7.2" | version_id) then "libsodium-dev" else empty end ),
				# oniguruma is part of mbstring in php 7.4+
				( if (.version | version_id) >= ("7.4" | version_id) then "libonig-dev" else empty end )
			end )
		] | sort[] | (
-}}
		{{ . }} \
{{
		)
	] | add
-}}
	; \
```

Reply:
hmm, I prefer the former (I've been moving the versioned bits to the end too, if they include a comment that otherwise breaks up the package list), but I defer to your preference :smile: